src/module-server/node-resolve.test.ts

```diff
@@ -0,0 +1,222 @@
+import { promises as fs } from 'fs';
+import { tmpdir } from 'os';
+import { dirname, join, normalize, posix, sep } from 'path';
+import { changeErrorMessage } from '../utils';
+import { nodeResolve } from './node-resolve';
+
+const createdPaths: string[] = [];
+
+// Cleanup created files
+afterAll(async () => {
+  await Promise.all(
+    createdPaths.map(async (path) => {
+      // For node 12, fs.rm() is not supported.
+      // For node 16, fs.rmdir() with recursive is deprecated (but still works)
+      // When we drop support for node 12, switch to below line for fs.rm()
+      await fs.rmdir(path, { recursive: true });
+      // Await fs.rm(path, { recursive: true, force: true });
+    }),
+  );
+});
+
+const createFs = async (input: string) => {
+  const dir = await fs.mkdtemp(join(tmpdir(), 'pleasantest-create-fs-'));
+  createdPaths.push(dir);
+  const paths = input
+    .trim()
+    .split('\n')
+    .map((line) => {
+      const [, originalPath, contents] = /(\S*)(.*)/.exec(line.trim()) || [];
+      const path = normalize(
+        join(dir, originalPath.split(posix.sep).join(sep)),
+      );
+      return { path, contents };
+    });
+
+  await Promise.all(
+    paths.map(async ({ path, contents }) => {
+      await fs.mkdir(dirname(path), { recursive: true });
+      await fs.writeFile(path, contents || '');
+    }),
+  );
+
+  /** Replaces all instances of randomized tmp dir with "." */
+  const unrandomizePath = (text: string) => text.split(dir).join('.');
+
+  const resolve = async (id: string, { from }: { from?: string } = {}) => {
+    const result = await nodeResolve(
+      id,
+      join(dir, from || 'index.js'),
+      dir,
+    ).catch((error) => {
+      throw changeErrorMessage(error, (error) => unrandomizePath(error));
+    });
+    if (result)
+      return unrandomizePath(typeof result === 'string' ? result : result.path);
+  };
+
+  return { resolve };
+};
+
+describe('resolving in node_modules', () => {
+  test('resolves main field', async () => {
+    const fs = await createFs(`
+      ./node_modules/foo/package.json {"main": "entry.js"}
+      ./node_modules/foo/entry.js
+    `);
+    expect(await fs.resolve('foo')).toBe('./node_modules/foo/entry.js');
```

Comment:
What I've done in the past for something like this is to make real files/folders for each test, but I like this better because it is easier to see the whole test in one place, and because changes are easier to make and will show up better in diffs, etc.